fix(KONFLUX-9792): fix rh-sign-image for single-container docker manifests

lhh opened the pull request:
## Describe your changes

This makes the `rh-sign-image` task match `rh-sign-image-cosign` with respect to looking for child manifests only in list-type mediatypes.

Without this fix, the `rh-sign-image` task was only accepting single-container OCI manifests and assuming all others were list-type manifests, which was causing failures whenever the BUILDAH_FORMAT was set to `docker` for single containers.

## Relevant Jira

KONFLUX-9792

## Checklist before requesting a review
- [N/A] I have marked as draft or added `do not merge` label if there's a dependency PR
  - If you want reviews on your draft PR, you can add reviewers or add the `release-service-maintainers` handle if you are unsure who to tag
- [x] My commit message includes `Signed-off-by: My name <email>`
- [x] I read CONTRIBUTING.MD and [commit formatting](CONTRIBUTING.md#commit-message-formatting-and-standards)
- [x] I have run the README.md generator script in `.github/scripts/readme_generator.sh` and verified the results using `.github/scripts/check_readme.sh`

lhh committed:
fix(KONFLUX-9792): fix rh-sign-image for docker container manifests

The Docker image format wasn't being accounted for within the
rh-sign-image task due to the task treating anything that was not
a single OCI image mediatype as a list-type mediatype.  This makes
the rh-sign-image task accept the docker mediatype as a non-list
as well.

Signed-off-by: Lon Hohberger <lhh@redhat.com>

arewm commented:
/ok-to-test

arewm commented on tasks/managed/rh-sign-image/rh-sign-image.yaml:
```diff
@@ -247,8 +247,10 @@ spec:
+            # For multi arch and index images, also sign all the manifests inside
+            MEDIATYPE="$(jq -r '.mediaType' <<< "$RAW_OUTPUT")"
+            if [ "$MEDIATYPE" == "application/vnd.oci.image.index.v1+json" ] || \
+               [ "$MEDIATYPE" == "application/vnd.docker.distribution.manifest.list.v2+json" ]; then
```
Do we have any test cases for other situations where we check for the Docker v2s2 media types?

mmalina replied:
Presumably this is the only place they saw an issue? Given that the rh-sign-cosign task does indeed check both of these, I think it should be fine to make this change.

mmalina replied:
And also, I looked if it makes sense to add a test for this. We do have a test for signing a multiarch image (`tasks/managed/rh-sign-image/tests/test-rh-sign-image-single-component-multi-arch.yaml`). Adding coverage for this would likely mean duplicating that test, just with the other media type string. Probably not much value in that.

lhh replied:
We haven't gotten past this, so I can't confirm nor deny whether other, later steps in the release pipeline would fail.

lhh replied:
Just so we're clear - this _isn't_ a multi-arch/index image. I saw single-instance for OCI, but not Docker (`test-rh-sign-image-single-component.yaml`).

lhh replied:
Here's what the code was running against:

```
{
  "schemaVersion": 2,
  "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
  "config": {
    "mediaType": "application/vnd.docker.container.image.v1+json",
    "size": 34835,
    "digest": "sha256:ec00ee53ba9645b1a40eb1bfe7420f9e2bd1b0b532a37237bed5c3a8eb89d550"
  },
  "layers": [
    {
      "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
      "size": 79977606,
      "digest": "sha256:1e0d3859902320579031b75c16b7945d5a4ddafd9fae85b6f7f1128659f6f26e"
    },
    {
      "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
      "size": 140891286,
      "digest": "sha256:dd85c74d1fe73af400c04e0a9d8e1cd002c6ee532afead97f2620c67c15ec2e9"
    },
    {
      "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
      "size": 30676092,
      "digest": "sha256:34245d8761530577f1516e7b6264788c6dd1aae3c3d94e4c231d18157bd80510"
    },
    {
      "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
      "size": 27232395,
      "digest": "sha256:3f9ee708b8241757f09522f2a92484f8781bcff14ddbbc65a75241c629020bda"
    },
    {
      "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
      "size": 8886287,
      "digest": "sha256:f8266280a85062784021fc530598c6c9e3787792e3e118a2056213b8d31bc689"
    }
  ]
}
```

The specific error without the patch is:
```
++ jq -r .mediaType
+ '[' application/vnd.docker.distribution.manifest.v2+json '!=' application/vnd.oci.image.manifest.v1+json ']'
++ jq -r '[.manifests[].digest] | join(" ")'
jq: error (at <stdin>:1): Cannot iterate over null (null)
```

mmalina replied:
One thought: Would it make it potentially safer to just add your type to the list of non-multi arch types, instead of switching the logic to list the multi arch types?
Meaning:
```
            if [ "$MEDIATYPE" != "application/vnd.oci.image.manifest.v1+json" ] && \
               [ "$MEDIATYPE" != "application/vnd.docker.distribution.manifest.v2+json" ]; then
```

mmalina replied:
Yeah, at first I read it wrong. But I would still prefer to just add the missing single type to that condition.

lhh replied:
@mmalina ack - I'll reverse the logic. The reason I picked using the list types was:

1. it matches the behavior of `rh-sign-image-cosign` task, and
2. the only types which will have a `.manifests[]` are list types - so if another non-list type hits this code, it will fail in the same way. That said, I don't know what other types are reasonable to expect besides the two we care about as above, and it's more clear the way you suggest.

lhh replied:
Updated. Thanks for the review!

mmalina replied:
Yeah, I don't really have a strong opinion on this, it just felt like less change = less chance of breaking some use case. But I think you're right that changing to the list types would actually likely reduce the risk of this breaking for something... Anyway, the current version looks ok.

lhh committed:
fix(KONFLUX-9792): add test for single docker images

Rather than adding a new single-component test for Docker
format images, add one to our existing multi-component
snapshot test.

Signed-off-by: Lon Hohberger <lhh@redhat.com>

lhh commented:
/ok-to-test

lhh commented:
/ok-to-test

arewm commented:
/ok-to-test

lhh commented:
I'm not sure why those tests failed on the remote hosts; I can't log into them to view the results to sort it out.

scoheb commented:
> I'm not sure why those tests failed on the remote hosts; I can't log into them to view the results to sort it out.

transient issue on that testing cluster. I restarted the 2 failed ones.

mmalina commented:
/ok-to-test

mmalina commented:
Collector test failed again. `Timeout waiting for PipelineRun to appear after 300 seconds`
